torchao/prototype/quantized_training/int8.py

```diff
@@ -0,0 +1,247 @@
+    out = Int8QTLinearWeight(
+        args[0].int_data.to(device=device),
+        args[0].scale.to(device=device, dtype=dtype),
+    )
+    return return_and_correct_aliasing(func, args, kwargs, out)
+
+
+# to make training work with existing PyTorch optimizers, we return a normal tensor
+@Int8QTLinearWeight.implements(aten.zeros_like.default)
+def _(func, types, args, kwargs):
+    dtype = kwargs.get("dtype", args[0].dtype)
+    device = kwargs.get("device", args[0].device)
+    return torch.zeros(args[0].shape, dtype=dtype, device=device)
+
+
+# out-of-place math ops always return plain tensor
+@Int8QTLinearWeight.implements([aten.sub.Tensor, aten.mul.Tensor])
+def _(func, types, args, kwargs):
+    args = [x.dequantize() if isinstance(x, Int8QTLinearWeight) else x for x in args]
+    return func(*args, **kwargs)
+
+
+@Int8QTLinearWeight.implements(aten.copy_.default)
+def _(func, types, args, kwargs):
+    if isinstance(args[0], Int8QTLinearWeight) and isinstance(args[1], Int8QTLinearWeight):
+        args[0].int_data.copy_(args[1].int_data, **kwargs)
+        args[0].scale.copy_(args[1].scale, **kwargs)
+
+    elif isinstance(args[0], Int8QTLinearWeight):
+        int_data, scale = Int8QTLinearWeight.quantize(args[1], stochastic_rounding=True)
+        args[0].int_data.copy_(int_data, **kwargs)
+        args[0].scale.copy_(scale, **kwargs)
+
+    else:
+        args[0].copy_(args[1].dequantize(), **kwargs)
+
+    return args[0]
+
+
+@Int8QTLinearWeight.implements([aten.addcdiv_.default, aten.add_.Tensor])
+def _(func, types, args, kwargs):
+    original = args[0]
+    out = func(args[0].dequantize(), *args[1:], **kwargs)
+    return original.copy_(out)
+
+
+# FSDP ops
+@Int8QTLinearWeight.implements(aten.split.Tensor)
+def _(func, types, args, kwargs):
+    if len(args) == 3 and args[2] != 0:
+        raise NotImplementedError("Int8QTLinearWeight only supports split at dim=0")
+
+    int8_weight: Int8QTLinearWeight = args[0]
+    int_data_list = func(int8_weight.int_data, *args[1:], **kwargs)
+    scale_list = func(int8_weight.scale, *args[1:], **kwargs)
+
+    out = [Int8QTLinearWeight(int_data, scale) for int_data, scale in zip(int_data_list, scale_list)]
+    return out
+
+
+@Int8QTLinearWeight.implements(aten.new_zeros.default)
+def _(func, types, args, kwargs):
+    size = args[1]
+    if len(size) != 2:
+        raise NotImplementedError
+
+    # TODO: handle pin_memory kwarg?
+    device = kwargs.get("device", args[0].device)
+    dtype = kwargs.get("dtype", args[0].dtype)
+    int_data = torch.zeros(size, device=device, dtype=torch.int8)
+    scale = torch.zeros(size[0], device=device, dtype=dtype)
+    return Int8QTLinearWeight(int_data, scale)
+
+
+# don't do anything. workaround for FSDP2. might give unexpected or wrong results.
+@Int8QTLinearWeight.implements([aten.view.default, aten.as_strided.default])
+def _(func, types, args, kwargs):
+    out = Int8QTLinearWeight(args[0].int_data, args[0].scale)
```

Comment:
Can you explain why we don't need to do anything here? What happens if we call view on the inner tensors and return a copy?

Reply:
These ops are called by FSDP2 by these lines (there are a few mores, you can ctrl+F, but these are the first hits when FSDP2 runs)

https://github.com/pytorch/pytorch/blob/cd565bc45554f6c7d5659bff19a270bda58ce71d/torch/distributed/_composable/fsdp/_fsdp_param.py#L349
https://github.com/pytorch/pytorch/blob/cd565bc45554f6c7d5659bff19a270bda58ce71d/torch/distributed/_composable/fsdp/_fsdp_param.py#L466-L471

From what I understand, FSDP2 expects a "view" here i.e. not a copy. Since we do channel-wise quantization here, a view doesn't quite make sense since it changes the dimensions -> quantization scale should change too (i.e. group size is tied to tensor dims). Propagate the ops to inner tensor may not make much sense too. For example, `as_strided` can't be called on `.scale`, since scale has different shape from the outer tensor.

I looked into NF4 impl, and it seems like they also keep the inner tensors as is, only update the kwargs passing to `Tensor._make_wrapper_subclass()`, which change the outer tensor appearance.

https://github.com/pytorch/ao/blob/e7fc0ed0f2e130e44d250541aba71c7a7a985f15/torchao/dtypes/nf4tensor.py#L230-L270

So basically, it's only a hack to make it work with FSDP2. Normally we shouldn't call these ops at all! I don't know if passing the "expected kwargs" to `Tensor._make_wrapper_subclass()` is necessary. The FSDP2 test passes on my machine (with an earlier version of pytorch nightly), so I think it might be ok? This is why I also request review from @awgu to check what is the best practice / recommended way here.

Reply:
Sounds good, can you add a comment in the code to explain this?